File: app/src/main/kotlin/com/app/backpackr/dagger/modules/NetworkModule.java
package com.app.backpackr.dagger.modules;

import com.squareup.moshi.Moshi;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.converter.moshi.MoshiConverterFactory;

/**
 * Created by kmikhailovskiy on 24.11.2016.
 */

@Module
public class NetworkModule {
    private String baseUrl;

    public NetworkModule(String baseUrl) {
        this.baseUrl = baseUrl;
    }

    @Provides
    @Singleton
    public Moshi provideMoshi() {
        return new Moshi.Builder().build();
    }

    @Provides
    @Singleton
    public OkHttpClient provideOkHttp() {
        return new OkHttpClient.Builder().build();
    }

    @Provides
    @Singleton
    public Retrofit provideRetrofit(Moshi moshi, OkHttpClient okHttpClient) {
        return new Retrofit.Builder()
                .addConverterFactory(MoshiConverterFactory.create())
                .baseUrl(baseUrl)
                .client(okHttpClient)
                .build();
    }
}

File: app/src/main/kotlin/com/app/backpackr/dagger/modules/NetworkModule.kt
package com.app.backpackr.dagger.modules

import com.app.backpackr.network.ApiBaseUrls
import com.app.backpackr.network.ApiHeaders
import com.app.backpackr.network.LoggingInterceptor
import com.app.backpackr.network.models.BaseUrl
import com.squareup.moshi.Moshi

import javax.inject.Singleton

import dagger.Module
import dagger.Provides
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory
import retrofit2.converter.moshi.MoshiConverterFactory
import java.util.concurrent.TimeUnit

/**
 * Created by kmikhailovskiy on 24.11.2016.
 */

@Module
class NetworkModule {
    val CONNECT_TIMEOUT: Long = 30

    @Provides
    @Singleton
    fun provideMoshi(): Moshi {
        return Moshi.Builder().build()
    }

    @Provides
    @Singleton
    fun provideBaseUrl(): BaseUrl {
        return BaseUrl(ApiBaseUrls.GOOGLE_PLACES.toString())
    }

    @Provides
    @Singleton
    fun provideOkHttpClient(): OkHttpClient {
        val builder: OkHttpClient.Builder = OkHttpClient.Builder()
                .connectTimeout(CONNECT_TIMEOUT, TimeUnit.SECONDS)
                .readTimeout(CONNECT_TIMEOUT, TimeUnit.SECONDS)
                .addInterceptor(LoggingInterceptor())

        return builder.build()
    }

    @Provides
    @Singleton
    fun provideRetrofit(baseUrl: BaseUrl, okHttpClient: OkHttpClient): Retrofit {
        return Retrofit.Builder()
                .addConverterFactory(MoshiConverterFactory.create())
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .baseUrl(baseUrl.endPointUrl)
                .client(okHttpClient)
                .build()
    }
}
